### DISCUSSION TITLE
Changes in scpiParser_parseAllProgramData

### DISCUSSION THREAD

**Discussion by Divi95** [j123b567/scpi-parser#124] [Ideas]
currently I am getting pContext->parser_state.programData.type as SCPI_TOKEN_ALL_PROGRAM_DATA
but in parser. c
int scpiParser_parseAllProgramData(lex_state_t * state, scpi_token_t * token, int * numberOfParameters)

line 1404
if we add this line token->type = tmp.type;

after adding the above statement (token->type = tmp.type), I can get the type of argument I have passed (for example if I have passed hex argument pContext->parser_state.programData.type will be updated bySCPI_TOKEN_HEXNUM

will this break something?

**Comment by j123b567** [comment]
I'm not an expert in parser/lexer theory, so there are some decisions in the code, that are not common and/or fully incorrect in sense of separation of lexer and parser.

One such thing is `scpiParser_parseAllProgramData` which is on parser side but generates completely new lexer token which represents all program data (EBNF grammar `(WS programData (COMMA programData)*)?`). Only one purpose of this function is to identify the end of the command, so yes, it will break everything, if you change the token type. This function is just for internal purpose and it should not be possible to call it outside the library.

If you need to process program data parameters one by one, then use `SCPI_Parameter` and friends. Which does the same as `scpiParser_parseAllProgramData` but it tells you correct lexer type for every parameter (formally programData), it finds. You can easilly iterate over all parameters. If  you know the type of the parameter in advance, you shuld call other functions, so it will cause a parser error if you expect int, but parameter is e.g. string and you don't need to handle all cases yourself.
